File: 백준/Gold/34464. Grid and Numbers Game/Grid and Numbers Game.cc
#include<bits/stdc++.h>
using namespace std;
using ll = long long;
using ai3 = array<int, 3>;
const int dx[] = {1, -1, 0, 0}, dy[] = {0, 0, 1, -1};

int main(){
    cin.tie(0); cout.tie(0); ios::sync_with_stdio(false);
    int t; cin >> t;
    while(t--){
        int n, m; cin >> n >> m;
        vector<vector<int>> d(n+2, vector<int>(m+2, -1));
        vector<ai3> v;
        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= m; j++){
                int x; cin >> x;
                v.push_back({x, i, j});
            }
        }
        sort(v.begin(), v.end());
        for(auto& [x, i, j] : v){
            for(int dir = 0; dir < 4; dir++) d[i][j] = max(d[i][j], d[i+dy[dir]][j+dx[dir]]+1);
        }
        int ans = 0;
        for(auto& [x, i, j] : v) ans ^= (x-d[i][j]);
        cout << (ans&1 ? "Yes\n" : "No\n");
    }
    return 0;
}

File: 백준/Gold/34464. Grid and Numbers Game/Grid and Numbers Game.py
import sys
def print(*args, sep=" ", end="\n"): return sys.stdout.write(sep.join(map(str, args)) + end)
def input(): return sys.stdin.readline().rstrip()

dx = (1, -1, 0, 0); dy = (0, 0, 1, -1)

def main():
    t = int(input())
    for _ in range(t):
        n, m = map(int, input().split())
        d = [[-1]*(m+2) for _ in range(n+2)]
        v = []
        for i in range(1, n+1):
            a = tuple(map(int, input().split()))
            for j in range(1, m+1):
                v.append((a[j-1], i, j))
        v.sort()
        for _, i, j in v:
            for dir in range(4): d[i][j] = max(d[i][j], d[i+dy[dir]][j+dx[dir]]+1)
        ans = 0
        for x, i, j in v: ans ^= (x-d[i][j])
        print('Yes' if ans&1 else 'No')
    return 0
if __name__ == '__main__':
    sys.exit(main())